sdk/tables/azure-data-tables/azure/data/tables/_encoder.py

```diff
@@ -0,0 +1,105 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import abc, enum
+from typing import Any, Optional, Tuple, Union, Mapping, TypeVar, Generic, Dict
+from uuid import UUID
+from datetime import datetime
+from math import isnan
+
+from ._entity import EdmType
+from ._entity import TableEntity
+from ._deserialize import _convert_to_entity
+from ._common_conversion import _encode_base64, _to_utc_datetime
+
+_ODATA_SUFFIX = "@odata.type"
+T = TypeVar("T")
+
+
+class TableEntityEncoderABC(abc.ABC, Generic[T]):
+    def prepare_key(self, key: str) -> str:
+        """Duplicate the single quote char to escape."""
+        try:
+            return key.replace("'", "''")
+        except (AttributeError, TypeError):
+            raise TypeError("PartitionKey or RowKey must be of type string.")
+
+    def prepare_value(
+        self, name: Optional[str], value: Any
+    ) -> Tuple[Optional[EdmType], Optional[Union[str, int, float, bool]]]:
+        """This is a utility to allow customers to reproduce our internal encoding for types like datetime, bytes, float and int64."""
+        if isinstance(value, bool):
+            return None, value
+        if isinstance(value, enum.Enum):
+            return self.prepare_value(name, value.value)
+        if isinstance(value, str):
+            return None, value
+        if isinstance(value, int):
+            return None, value  # TODO: Test what happens if the supplied value exceeds int32.
+        if isinstance(value, float):
+            if isnan(value):
+                return EdmType.DOUBLE, "NaN"
+            if value == float("inf"):
+                return EdmType.DOUBLE, "Infinity"
+            if value == float("-inf"):
+                return EdmType.DOUBLE, "-Infinity"
+            return EdmType.DOUBLE, value
+        if isinstance(value, UUID):
+            return EdmType.GUID, str(value)
+        if isinstance(value, bytes):
+            return EdmType.BINARY, _encode_base64(value)
+        if isinstance(value, datetime):
+            try:
+                if value.tables_service_value:
+                    return EdmType.DATETIME, value.tables_service_value
+            except AttributeError:
+                pass
+            return EdmType.DATETIME, _to_utc_datetime(value)
+        if isinstance(value, tuple):
+            if value[1] == EdmType.INT64:  # TODO: Test if this works with either string or enum input.
+                return EdmType.INT64, str(value[0])  # TODO: Test what happens if the supplied value exceeds int64
+            _, encoded_value = self.prepare_value(name, value[0])
```

Comment:
For full backcompat - I guess we'll need to fully expand this out.
We should probably move this into a separate private function.
```suggestion
            edm_type = value[1]
            unencoded_value = value[0]
            if unencoded_value is None:
                return edm_type, value
            if edm_type == EdmType.STRING:
                return EdmType.STRING, str(unencoded_value)
            if edm_type == EdmType.INT64:
                return EdmType.INT64, str(unencoded_value) 
            if edm_type == EdmType.INT32:
                return EdmType.INT32, int(unencoded_value)
            if edm_type == EdmType.BOOLEAN:
                return EdmType.BOOLEAN, unencoded_value  # Leaving unaltered as per original implementation.
            if edm_type == EdmType.GUID:
                return EdmType.GUID, str(unencoded_value)
            if edm_type == EdmType.BINARY:
                return EdmType.BINARY, _encode_base64(unencoded_value)  # Leaving this with the double-encoding bug for now, as per original implementation
            if edm_type == EdmType.DOUBLE:
                if isinstance(value, str):
                    # Pass a serialized value straight through
                    return EdmType.DOUBLE, value
                if isnan(value):
                    return EdmType.DOUBLE, "NaN"
                if value == float("inf"):
                    return EdmType.DOUBLE, "Infinity"
                if value == float("-inf"):
                    return EdmType.DOUBLE, "-Infinity"
                return EdmType.DOUBLE, value
            if edm_type == EdmType.DATETIME:
                if isinstance(value, str):
                    # Pass a serialized datetime straight through
                    return EdmType.DATETIME, value
                try:
                    # Check is this is a 'round-trip' datetime, and if so
                    # pass through the original value.
                    if value.tables_service_value:
                        return EdmType.DATETIME, value.tables_service_value
                except AttributeError:
                    pass
                return EdmType.DATETIME, _to_utc_datetime(value)
            
```

Reply:
Updated. I added to throw a `TypeError` when the edm type is not in any of the `if` check.